file/file.go

```diff
@@ -0,0 +1,166 @@
+	rdr       io.Reader
+}
+
+var _ ipld.Node = (*shardNodeFile)(nil)
+
+func (s *shardNodeFile) Read(p []byte) (int, error) {
+	// collect the sub-nodes on first use
+	if s.rdr == nil {
+		links, err := s.substrate.LookupByString("Links")
+		if err != nil {
+			return 0, err
+		}
+		readers := make([]io.Reader, 0)
+		lnki := links.ListIterator()
+		for !lnki.Done() {
+			_, lnk, err := lnki.Next()
+			if err != nil {
+				return 0, err
+			}
+			if pbl, ok := lnk.(dagpb.PBLink); ok {
+				target, err := s.lsys.Load(ipld.LinkContext{Ctx: s.ctx}, pbl.Hash.Link(), basicnode.Prototype.Any)
+				if err != nil {
+					return 0, err
+				}
+
+				asFSNode, err := NewUnixFSFile(s.ctx, target, s.lsys)
+				if err != nil {
+					return 0, err
+				}
+				readers = append(readers, asFSNode)
+			} else {
+				return 0, fmt.Errorf("unsupported link type: %T", lnk)
+			}
+		}
+		s.rdr = io.MultiReader(readers...)
+	}
+	return s.rdr.Read(p)
+}
+
+func (s *shardNodeFile) Kind() ipld.Kind {
+	return ipld.Kind_Bytes
+}
+
+func (s *shardNodeFile) AsBytes() ([]byte, error) {
+	return io.ReadAll(s)
+}
+
+func (s *shardNodeFile) AsBool() (bool, error) {
+	return false, ipld.ErrWrongKind{TypeName: "bool", MethodName: "AsBool", AppropriateKind: ipld.KindSet_JustBytes}
+}
+
+func (s *shardNodeFile) AsInt() (int64, error) {
+	return 0, ipld.ErrWrongKind{TypeName: "int", MethodName: "AsInt", AppropriateKind: ipld.KindSet_JustBytes}
+}
+
+func (s *shardNodeFile) AsFloat() (float64, error) {
+	return 0, ipld.ErrWrongKind{TypeName: "float", MethodName: "AsFloat", AppropriateKind: ipld.KindSet_JustBytes}
+}
+
+func (s *shardNodeFile) AsString() (string, error) {
+	return "", ipld.ErrWrongKind{TypeName: "string", MethodName: "AsString", AppropriateKind: ipld.KindSet_JustBytes}
+}
+
+func (s *shardNodeFile) AsLink() (ipld.Link, error) {
+	return nil, ipld.ErrWrongKind{TypeName: "link", MethodName: "AsLink", AppropriateKind: ipld.KindSet_JustBytes}
+}
+
+func (s *shardNodeFile) AsNode() (ipld.Node, error) {
+	return nil, nil
+}
+
+func (s *shardNodeFile) Size() int {
+	return 0
+}
+
+func (s *shardNodeFile) IsAbsent() bool {
+	return false
+}
+
+func (s *shardNodeFile) IsNull() bool {
+	return s.substrate.IsNull()
+}
+
+func (s *shardNodeFile) Length() int64 {
+	return 0
+}
+
+func (s *shardNodeFile) ListIterator() ipld.ListIterator {
+	return nil
+}
+
+func (s *shardNodeFile) MapIterator() ipld.MapIterator {
+	return nil
+}
+
+func (s *shardNodeFile) LookupByIndex(idx int64) (ipld.Node, error) {
+	return nil, ipld.ErrWrongKind{}
+}
+
+func (s *shardNodeFile) LookupByString(key string) (ipld.Node, error) {
+	return nil, ipld.ErrWrongKind{}
+}
+
+func (s *shardNodeFile) LookupByNode(key ipld.Node) (ipld.Node, error) {
+	return nil, ipld.ErrWrongKind{}
+}
+
+func (s *shardNodeFile) LookupBySegment(seg ipld.PathSegment) (ipld.Node, error) {
+	return nil, ipld.ErrWrongKind{}
+}
+
+func (s *shardNodeFile) Prototype() ipld.NodePrototype {
+	return basicnode.Prototype__Bytes{}
```

Comment:
This should return something more self-descriptive.

What exactly "self-descriptive" means here, I don't know -- it might be something that has to be somewhat best-effort.  Maybe it should peek at the root blocks of the thing and try to see if there's any sharding parameters, and return something with those if possible?  If it can't infer all the parameters that were used creating this ADL the first time, still returning a best guess that's at least the same ADL but with default parameters is probably better than nothing.

The overall purpose of `.Prototype()` is so that things like [traversal.FocusedTransform](https://pkg.go.dev/github.com/ipld/go-ipld-prime/traversal#Progress.FocusedTransform) (or the [traversal.TransformFn](https://pkg.go.dev/github.com/ipld/go-ipld-prime/traversal#TransformFn) that the user provides, if this is the leaf thing being replaced) can regenerate things correctly while making a copy-on-write update.  So if it forgets that something is an ADL we get very different behavior in a way that's probably not desired!